testdata/loader.go

```diff
@@ -0,0 +1,92 @@
+package testdata
+
+import (
+	"embed"
+	_ "embed"
+	"fmt"
+	"path/filepath"
+
+	"gopkg.in/launchdarkly/go-sdk-common.v2/ldvalue"
+)
+
+//go:embed data-files
+var dataFilesRoot embed.FS
+
+const dataBasePath = "data-files"
+
+// SourceInfo represents JSON or YAML data that was read from a file, after post-processing to expand
+// constants and parameters. For non-parameterized tests, you will get one SourceInfo per file. For
+// parameterized tests, there can be many instances per file, each with its own version of Data.
+// See ReadFile and docs/data_files.md.
+type SourceInfo struct {
+	FilePath string
+	BaseName string
+	Params   map[string]ldvalue.Value
+	Data     []byte
+}
+
+func (s SourceInfo) ParseInto(target interface{}) error {
+	if err := ParseJSONOrYAML(s.Data, target); err != nil {
+		return fmt.Errorf("error parsing %q %s: %w", s.BaseName, s.ParamsString(), err)
+	}
+	return nil
+}
+
+func (s SourceInfo) ParamsString() string {
+	if len(s.Params) == 0 {
+		return ""
+	}
+	ps := ""
+	for k, v := range s.Params {
+		if ps != "" {
+			ps += ","
+		}
+		ps += k + "=" + v.String()
+	}
+	return "(" + ps + ")"
+}
+
+// LoadDataFile reads a data file and performs any necessary constant/parameter substitutions. It can
+// return more than one SourceInfo because any file can be a parameterized test. See docs/data_files.md.
+//
+// The path parameter is relative to testdata/data-files.
+func LoadDataFile(path string) ([]SourceInfo, error) {
+	ret := make([]SourceInfo, 0, 10)
```

Comment:
No special reason. We know we'll have at least one element, so it wouldn't make sense to start with a zero-length backing array, but due to parameter expansion it's pretty likely that we'll end up with more than one. It will be expanded if necessary by `append`, so it makes very little difference except in terms of avoiding unnecessary extra reallocations... not that that is a significant concern in a test app like this.

Reply:
In that case, I'd just do a normal make w/o the parameter. I think the slight bit of ambiguity/confusion it raises is more than it's worth in terms of performance optimization, which may be premature anyways.

Reply:
Not sure what you mean by "premature"— if things are working as intended, we absolutely know there will be at least one element. While I'm fine with putting 1 instead of 10 here, it's not like I'm doing anything unusual here that might turn out to be more trouble than it's worth; preallocating array storage you are definitely going to use is a pretty normal to do in Go.

Reply:
Although really, it is virtually guaranteed to end up with more than 1 anyway, because the tests are only using data files like this in cases where we do want to use the parameter expansion feature.

I feel like "premature optimization" is a term that's sometimes thrown around a bit too easily. There is certainly a real problem that can happen if, based on a hypothetical scenario that might or might not ever really be relevant, design decisions are made that constrain you in some other way, or introduce unnecessary complexity, or have overhead that's not worthwhile. But this isn't an example of any of those. Preallocating a little leeway in a slice is common in Go, and slices in Go were designed so as to make that extremely transparent; I'm not seeing where the ambiguity or confusion would arise.

To be clear, I didn't write the line this way due to a real concern that the code would run too slowly otherwise. This isn't a performance-critical area and we're talking about trivial amounts of data. But it's fairly idiomatic Go and I think it's a good habit to use this pattern in general, because other code that we maintain _is_ performance-critical and there just isn't a great reason to use a no-parameter `make` for slices by default.

Reply:
What I think is a more realistic concern regarding `make` parameters is that it's unfortunately easy in Go to write `make([]x, 10)` when you meant `make([]x, 0, 10)` and end up setting the length rather than the capacity. But that's so rarely a thing that you want to do on purpose that I think it's best to just set up linting to warn you about any such usage, and then disable the linter for that line if you really meant to do it.

Reply:
By premature I mean do we know this is a critical path? Do we know this saves a meaningful amount of time? 

You mentioned above that it isn't a significant concern in a test app like this, so I would be in favor of leaving it out or picking `1`. 

Otherwise, it will be confusing when reading the source to know why 10 was chosen, or if it can be deleted.

I do agree it's not unusual to reserve some capacity, but it'd be great to preserve the intention of it in the form of a comment, or not raise questions at all by removing it.

Reply:
Actually, hold on— what are we even talking about? There's no such thing as zero-parameter `make` for a slice; `make([]SourceInfo)` is not a thing. You have to pick either a length, or a length and a capacity.

And if I set the capacity to 1, then I really don't think that's any different in terms of your concern of "knowing why it was chosen": a 1 with no explanation implies that we don't expect there to be more than 1, and that's not really the case. The answer to "why is this capacity chosen" is virtually always going to boil down to "I feel like this number is generally going to make sense in typical usage." Due to how slices work in Go, the capacity is never a hard limit.

Reply:
What we definitely do not want to ever do is `make([]SourceInfo, 0)`. There's literally no situation I can think of in which a zero-length slice with a zero-length backing array is advantageous.

Reply:
I've added a brief comment which I hope is enough, although I honestly feel like this is to some degree a common-sense thing and that the concern over it being confusing is excessive. I mean, this is a method whose description clearly says it can return multiple results. If I saw a method for, say, building a string of unknown length, and the initial buffer size the author picked was 100, it would be a bit odd for me to press the question of "why 100? why not 1 or 10?" unless I had a serious concern that this was inefficient. There's such a thing as premature _de_-optimization too.

Reply:
Ah, I think it's a personal preference for sure then. I certainly would ask "why 100?". I thought it was confusing, which is justification enough for me, but of course maybe someone else reviewing might not think so 🤷‍♂️

Reply:
Let me put it another way. There's nothing wrong with asking "why 100?" Often something isn't immediately obvious in a code review and it's fine to ask. But in a case like the one I just described— which I think is pretty analogous to this case— I think it's a bit much to say it is inherently confusing and ought to be changed to 1 or 0 for that reason, especially given that in Go you basically _never_ want it to be 0. Sometimes "because the point of this method is to build a list that's very likely to have more than 1 thing" is an adequate explanation and can be inferred from the basic fact of what the method is doing, and doesn't constitute premature optimization.